crates/volta-core/src/tool/serial.rs

```diff
@@ -83,30 +83,50 @@ impl Spec {
         Ok(tools)
     }
 
-    /// Check the args for the bad pattern of `volta install <tool> <number>`.
+    /// Check the args for the bad patterns of
+    /// - `volta install <number>`
+    /// - `volta install <tool> <number>`
     fn check_args<T>(args: &[T], action: &str) -> Fallible<()>
     where
         T: AsRef<str>,
     {
         let mut args = args.iter();
 
-        // The case we are concerned with is where we have `<tool> <number>`.
-        // This is only interesting if there are exactly two args. Then we care
-        // whether the two items are a bare name (with no `@version`), followed
-        // by a valid version specifier (ignoring custom tags). That is:
-        //
-        // - `volta install node@lts latest` is allowed.
-        // - `volta install node latest` is an error.
-        // - `volta install node latest yarn` is allowed.
-        if let (Some(name), Some(maybe_version), None) = (args.next(), args.next(), args.next()) {
-            if !HAS_VERSION.is_match(name.as_ref()) && is_version_like(maybe_version.as_ref()) {
-                return Err(ErrorKind::InvalidInvocation {
-                    action: action.to_string(),
-                    name: name.as_ref().to_string(),
-                    version: maybe_version.as_ref().to_string(),
+        match (args.next(), args.next(), args.next()) {
+            (Some(maybe_version), None, None) => {
```

Comment:
That's a great question! In my experience, it usually comes down to readability: I find the `<pattern> if <predicate> =>` can very easily get verbose enough that it's more difficult to read than `<pattern> => if <predicate>`. For your second question, yes if you use the predicate matching it would be more idiomatic to have the match provide the value directly.

In this case, I would probably lean that doing predicate matching is _slightly_ cleaner, but it's not anything that would block a PR. So if you'd like to (and have the time to) add another commit, that would be a bonus! But if not no worries and I'm happy to merge as is :+1:

Reply:
Thanks for the response! Pushed up a commit doing that.